### PR TITLE
Clean up `VolumeSnapshots` and `VolumeSnapshotContents`

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area ops-productivity
/kind enhancement

**What this PR does / why we need it**:
This PR enhances Gardener's clean-up logic to also include `VolumeSnapshots` and `VolumeSnapshotContents`. As opposed to other resources, those are only forcefully cleaned up after `1h` because deleting snapshots can take a while depending on the number of snapshots which were taken before.

However, `5m` (default) after the `VolumeSnapshotConent` has been deleted for the first time, an annotation `snapshot.storage.kubernetes.io/volumesnapshot-being-deleted=yes` is added to the affected object. It triggers a more or less forceful but graceful deletion through the CSI snapshotter, i.e. the snapshot in the infrastructure is deleted as well as the `VolumeSnapshotConent` if the previous step succeeded. The `snapshot.storage.kubernetes.io/volumesnapshot-being-created` annotation is removed likewise to not block the deletion flow.

**Release note**:

```feature operator
Gardener now cleans up `VolumeSnapshots` and `VolumeSnapshotContents` during cluster deletion. These resources are forcefully deleted after a `1h` grace period which eventually lead to leaked snapshots on the cloud provider side. Hence, if the CSI-Snapshotter cannot delete affected snapshots successfully for `1h`, operators/shoot-owners have to purge them manually.
```

### CODE REVIEW

File: pkg/operation/botanist/cleanup.go [hunk 1]
```diff
@@ -266,6 +274,10 @@ func (b *Botanist) CleanKubernetesResources(ctx context.Context) error {
 		cleanResourceFn(ops, c, &corev1.ServiceList{}, ServiceCleanOption, cleanOptions),
 		cleanResourceFn(ops, c, &appsv1.StatefulSetList{}, StatefulSetCleanOption, cleanOptions),
 		cleanResourceFn(ops, c, &corev1.PersistentVolumeClaimList{}, PersistentVolumeClaimCleanOption, cleanOptions),
+		// Cleaning up VolumeSnapshots can take a longer time if many snapshots were taken.
+		// Hence, we only finalize these objects after 1h.
+		cleanResourceFn(ops, c, &volumesnapshotv1beta1.VolumeSnapshotList{}, VolumeSnapshotContentCleanOption, FinalizeAfterOneHour),
+		cleanResourceFn(ops, c, &volumesnapshotv1beta1.VolumeSnapshotContentList{}, VolumeSnapshotContentCleanOption, FinalizeAfterOneHour),
```
Comment (rfranzke) [inline pkg/operation/botanist/cleanup.go]: Can we still respect the `cleanOptions` here, i.e., if somebody overwrote the default grace period with the `shoot.gardener.cloud/cleanup-kubernetes-resources-finalize-grace-period-seconds` annotation?

**Review by rfranzke — CHANGES_REQUESTED** [review]
Generally lgtm, thanks!
Still,
/needs second-opinion

**Comment by timuthy** [comment]
I added d14534c which implements the proposal mentioned in the PR description. It works perfectly for different scenarios, e.g.:
- Snapshot + Content are valid
- Snapshot is gone, Content is stuck in deletion

It's of course slightly implementation specific but still helps to eventually also delete the snapshots on the cloud provider side. Please let me know what you think about it.

PS: I'll add the missing unit tests for `volumeSnapshotContentCleaner` later if we decided to make use of it.

File: pkg/utils/kubernetes/client/client.go [hunk 2]
```diff
@@ -201,6 +206,57 @@ func (cl *cleaner) doClean(ctx context.Context, c client.Client, obj client.Obje
...
+	default:
+		return fmt.Errorf("type %T does neither implement client.Object nor client.ObjectList", obj)
+	}
+
+	return defaultCleaner.Clean(ctx, c, obj, opts...)
+}
+
+const annVolumeSnapshotBeingDeleted = "snapshot.storage.kubernetes.io/volumesnapshot-being-deleted"
+
+func (v volumeSnapshotContentCleaner) addDeleteAnnotation(ctx context.Context, c client.Client, obj client.Object, _ *CleanOptions) error {
```
Comment (MartinWeindel) [inline pkg/utils/kubernetes/client/client.go]: You may also want to remove the annotation `snapshot.storage.kubernetes.io/volumesnapshot-being-created` if it is set. See [comment in issue 1505](https://github.tools.sap/kubernetes-canary/issues-canary/issues/1505#issuecomment-471097) where setting the `volumesnapshot-being-deleted` was not sufficient.

**Comment by rfranzke** [comment]
I like the approach @timuthy, feel free to proceed as you planned :)

**Comment by timuthy** [comment]
Thanks for the feedback @rfranzke and @MartinWeindel. I adapted the PR so that you can have another look.
/remove author-action

File: pkg/utils/kubernetes/client/client.go [hunk 3]
```diff
@@ -201,6 +207,89 @@ func (cl *cleaner) doClean(ctx context.Context, c client.Client, obj client.Obje
...
+		return nil
+	}
+
+	patch := client.MergeFrom(obj.DeepCopyObject().(client.Object))
+
+	if obj.GetAnnotations() == nil {
+		obj.SetAnnotations(make(map[string]string))
+	}
+
+	// annVolumeSnapshotBeingDeleted triggers the CSI side-care to delete the snapshot on the infrastructure.
```
Comment (rfranzke) [inline pkg/utils/kubernetes/client/client.go]: Do you mean sidecar?

**Review by rfranzke — DISMISSED** [review]
/lgtm

File: pkg/utils/kubernetes/client/client.go (hunk 3 above)
Comment (timuthy) [reply pkg/utils/kubernetes/client/client.go]: Yes 😄

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**